### PR TITLE
Analytic wigner

### PR DESCRIPTION
**Pull request by JanosJiri**
Modifying wigner function with analytic formulas and different grid for numeric integration

### CODE REVIEW

**Comment by JanosJiri** [comment]
Analytic Wigner transform added for `gauss` and `sin` envelopes. I couldn't get Lorentzian analytic, `sin2` is giving me a complex formula, although it should be purely real. I will try to get the others analytic later but since we have the review now, we need to publish the code this week. 

The analytic formulas bring a massive speed up (we could possibly extend testing of pda and pdaw with Gaussian pulse to more samples). I also needed to modify tests.

@danielhollas have a look, please. I would like to merge this and then release one stable version for the manuscript. Anything we do after this week will not appear in the paper. Thanks!

File: supplementary/envelopes_analysis/envelope_wigner_transform.png
[the remark this answers is not in the document]
Comment (JanosJiri) [reply supplementary/envelopes_analysis/envelope_wigner_transform.png:1]: I see, I haven't thought about that. I will merge it with the new figure which should be consistent with the code but I don't expect it to be changed again. I will avoid replotting the figure again unless some major modifications to the envelopes happens. Thanks for pointing that out.

File: src/promdens/promdens.py [hunk 1]
```diff
@@ -172,34 +171,50 @@ def wigner_transform(self, tprime, de):
...
-        }
-
-        # instead of calculating the complex integral int_{-inf}^{inf}[E(t+s/2)E(t-s/2)exp(i(w-de)s)]ds we use the
-        # properties of even and odd fucntions and calculate 2*int_{0}^{inf}[E(t+s/2)E(t-s/2)cos((w-de)s)]ds
-        s = np.arange(0, factor[self.envelope_type]*self.fwhm, step=ds)
-        cos = np.cos((de - loc_omega)*s)
-        integral = 2*np.trapz(x=s, y=cos*self.calc_field_envelope(tprime + s/2)*self.calc_field_envelope(tprime - s/2))
+        # analytic integrals if available
+        if self.envelope_type == 'gauss':
+            integral = 16**(-(tprime - self.t0)**2/self.fwhm**2)*np.exp(-self.fwhm**2*(de - loc_omega)**2/np.log(16))*self.fwhm*np.sqrt(
```
Comment (danielhollas) [inline src/promdens/promdens.py:178]: To make this more readable, perhaps you could extract the prefactor to a separate variable

```suggestion
            prefactor = 16**(-(tprime - self.t0)**2/self.fwhm**2)*self.fwhm*np.pi/np.log(2)
            integral = prefactor * np.exp(-self.fwhm**2*(de - loc_omega)**2/np.log(16))
```

File: src/promdens/promdens.py [hunk 2]
```diff
@@ -172,34 +171,50 @@ def wigner_transform(self, tprime, de):
...
+                            np.pi*np.cos(np.pi*(tprime - self.t0)/self.fwhm)*np.sin(2*(tprime - self.t0 + self.fwhm)*w))/
+                            (w*(np.pi**2 - 4*self.fwhm**2*w**2)))
+            elif tprime >= self.t0 and tprime < self.t0 + self.fwhm:
+                integral = (np.pi*(2*self.fwhm*w*np.cos(2*(-tprime + self.t0 + self.fwhm)*w)*np.sin(np.pi*(tprime - self.t0)/self.fwhm) +
+                            np.pi*np.cos(np.pi*(tprime - self.t0)/self.fwhm)*np.sin(2*(-tprime + self.t0 + self.fwhm)*w))/
+                            (w*(np.pi**2 - 4*self.fwhm**2*w**2)))
+            else:
+                integral = 0
+        else:  # numerical integrals for the remaining envelopes
+            # setting an adaptive integration step according to the frequency of the integrand oscillations (de - omega)
```
Comment (danielhollas) [inline src/promdens/promdens.py:195]: Assuming that the numerical code did not change?

File: src/promdens/promdens.py [hunk 3]
```diff
@@ -172,34 +171,50 @@ def wigner_transform(self, tprime, de):
...
-            'sin2': 4,
-        }
-
-        # instead of calculating the complex integral int_{-inf}^{inf}[E(t+s/2)E(t-s/2)exp(i(w-de)s)]ds we use the
-        # properties of even and odd fucntions and calculate 2*int_{0}^{inf}[E(t+s/2)E(t-s/2)cos((w-de)s)]ds
-        s = np.arange(0, factor[self.envelope_type]*self.fwhm, step=ds)
-        cos = np.cos((de - loc_omega)*s)
-        integral = 2*np.trapz(x=s, y=cos*self.calc_field_envelope(tprime + s/2)*self.calc_field_envelope(tprime - s/2))
+        # analytic integrals if available
+        if self.envelope_type == 'gauss':
```
Comment (danielhollas) [inline src/promdens/promdens.py:177]: Just as a meta-comment: These extra if statements make me wonder more and more if we should have a LaserPulse base class and then subclass it for each LaserPulse type. Would make this a bit cleaner overall. Definitely not something urgent but I'll think about it a bit.

File: src/promdens/promdens.py [hunk 4]
```diff
@@ -172,34 +171,50 @@ def wigner_transform(self, tprime, de):
...
+            integral = 16**(-(tprime - self.t0)**2/self.fwhm**2)*np.exp(-self.fwhm**2*(de - loc_omega)**2/np.log(16))*self.fwhm*np.sqrt(
+                np.pi/np.log(2))
+        elif self.envelope_type == 'sin':
+            if de == loc_omega:
+                loc_omega += 1e-10  # because we can divide by 0 but the integral is constant for small omega
+            w = de - loc_omega
+            if tprime < self.t0 and tprime > self.t0 - self.fwhm:
```
Comment (danielhollas) [inline src/promdens/promdens.py:184]: I guess you could reference the `self.tmin` here? But perhaps this is better as it is more explicit

```suggestion
            if tprime < self.t0 and tprime > self.tmin:
```

```

**Review by danielhollas — APPROVED** [review]
This is an awesome work! Looks really non-trivial. It's great to see the speed improvements --- I don't think that is a given outcome. It could be that the analytical formula were so complicated that they would not be much faster in the end. :-) 

Nothing major just some nits.

Regarding testing, it would be interesting the see the errors (numerical versus analytical) as a function of tprime, since based on the number in the unit tests it sure seems like the errors will depend on tprime and might be different for sin and gauss waves. But this is perhaps more of a curiosity, I trust that you've tested this well.

File: tests/test_pulse.py [hunk 5]
```diff
@@ -347,8 +347,8 @@ def test_pulse_wigner_de_equals_omega(make_pulse, envelope_type):
             },
             "sin": {
                 0: 2.0,
-                1: 1.999019683371722,
-                2: 2.9995195653237154e-35,
+                1: 1.9990196998183594,
```
[the remark this answers is not in the document]
Comment (danielhollas) [reply tests/test_pulse.py:350]: Ah, it is hard coded in the code :-D

[the remark this answers is not in the document]
Comment (JanosJiri) [reply tests/test_pulse.py:350]: Yeah, I think so. It was decreasing when I was increasing the integration grid.

File: src/promdens/promdens.py [hunk 6]
```diff
@@ -172,34 +171,50 @@ def wigner_transform(self, tprime, de):
...
-        s = np.arange(0, factor[self.envelope_type]*self.fwhm, step=ds)
-        cos = np.cos((de - loc_omega)*s)
-        integral = 2*np.trapz(x=s, y=cos*self.calc_field_envelope(tprime + s/2)*self.calc_field_envelope(tprime - s/2))
+        # analytic integrals if available
+        if self.envelope_type == 'gauss':
+            integral = 16**(-(tprime - self.t0)**2/self.fwhm**2)*np.exp(-self.fwhm**2*(de - loc_omega)**2/np.log(16))*self.fwhm*np.sqrt(
+                np.pi/np.log(2))
+        elif self.envelope_type == 'sin':
+            if de == loc_omega:
+                loc_omega += 1e-10  # because we can divide by 0 but the integral is constant for small omega
```
[the remark this answers is not in the document]
Comment (JanosJiri) [reply src/promdens/promdens.py:182]: Aaah, yeah, you are right. I will change it.

File: tests/test_pulse.py [hunk 7]
```diff
@@ -347,8 +347,8 @@ def test_pulse_wigner_de_equals_omega(make_pulse, envelope_type):
...
+                2: 0.0,
```
[the remark this answers is not in the document]
Comment (danielhollas) [reply tests/test_pulse.py:351]: Ah, it is hard coded in the code :-D

File: src/promdens/promdens.py (hunk 1 above)
Comment (JanosJiri) [reply src/promdens/promdens.py:178]: Yes, that would make sense.

File: src/promdens/promdens.py (hunk 2 above)
Comment (JanosJiri) [reply src/promdens/promdens.py:195]: No, nothing changed.

**Pull request merged**